scripts/maintenance/computational-clusters/autoscaled_monitor/ec2.py

```diff
@@ -0,0 +1,171 @@
+                for key, value in custom_tags.items()
+            ]
+        )
+
+    if user_id:
+        ec2_filters.append({"Name": "tag:user_id", "Values": [f"{user_id}"]})
+    if wallet_id:
+        ec2_filters.append({"Name": "tag:wallet_id", "Values": [f"{wallet_id}"]})
+
+    return ec2_resource.instances.filter(Filters=ec2_filters)
+
+
+async def list_computational_instances_from_ec2(
+    state: AppState,
+    user_id: int | None,
+    wallet_id: int | None,
+) -> ServiceResourceInstancesCollection:
+    assert state.environment["PRIMARY_EC2_INSTANCES_KEY_NAME"]
+    assert state.environment["WORKERS_EC2_INSTANCES_KEY_NAME"]
+    assert (
+        state.environment["PRIMARY_EC2_INSTANCES_KEY_NAME"]
+        == state.environment["WORKERS_EC2_INSTANCES_KEY_NAME"]
+    ), "key name is different on primary and workers. TIP: adjust this code now"
+    custom_tags = {}
+    if state.environment["PRIMARY_EC2_INSTANCES_CUSTOM_TAGS"]:
+        assert (
+            state.environment["PRIMARY_EC2_INSTANCES_CUSTOM_TAGS"]
+            == state.environment["WORKERS_EC2_INSTANCES_CUSTOM_TAGS"]
+        ), "custom tags are different on primary and workers. TIP: adjust this code now"
+        custom_tags = json.loads(state.environment["PRIMARY_EC2_INSTANCES_CUSTOM_TAGS"])
+    assert state.ec2_resource_clusters_keeper
+    return await _list_running_ec2_instances(
+        state.ec2_resource_clusters_keeper,
+        state.environment["PRIMARY_EC2_INSTANCES_KEY_NAME"],
+        custom_tags,
+        user_id,
+        wallet_id,
+    )
+
+
+async def list_dynamic_instances_from_ec2(
+    state: AppState,
+    user_id: int | None,
+    wallet_id: int | None,
+) -> ServiceResourceInstancesCollection:
+    assert state.environment["EC2_INSTANCES_KEY_NAME"]
+    custom_tags = {}
+    if state.environment["EC2_INSTANCES_CUSTOM_TAGS"]:
+        custom_tags = json.loads(state.environment["EC2_INSTANCES_CUSTOM_TAGS"])
+    assert state.ec2_resource_autoscaling
+    return await _list_running_ec2_instances(
+        state.ec2_resource_autoscaling,
+        state.environment["EC2_INSTANCES_KEY_NAME"],
+        custom_tags,
+        user_id,
+        wallet_id,
+    )
+
+
+_DEFAULT_BASTION_NAME: Final[str] = "bastion-host"
+
+
+@cached()
+async def get_computational_bastion_instance(state: AppState) -> Instance:
+    assert state.ec2_resource_clusters_keeper  # nosec
+    assert state.environment["PRIMARY_EC2_INSTANCES_KEY_NAME"]  # nosec
+    instances = await _list_running_ec2_instances(
+        state.ec2_resource_clusters_keeper,
+        state.environment["PRIMARY_EC2_INSTANCES_KEY_NAME"],
+        {},
+        None,
+        None,
+    )
+
+    possible_bastions = list(
+        filter(lambda i: _DEFAULT_BASTION_NAME in get_instance_name(i), instances)
+    )
+    assert len(possible_bastions) == 1
+    return possible_bastions[0]
+
+
+@cached()
+async def get_dynamic_bastion_instance(state: AppState) -> Instance:
```

Comment:
I wonder why wouldn't you use terraform outputs that give you IP of the bastion

Reply:
@YuryHrytsuk  I did not find it in osparc-staging.io?

Reply:
all clear, so until I actually start reading these terraform files I will keep it so. very nice thanks for the clarifications.